core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -220,6 +221,7 @@ public void serverShutdown() {
       synchronized (ServerImpl.this) {
         // transports collection can be modified during shutdown(), even if we hold the lock, due
         // to reentrancy.
+        transportServerTerminated = true;
```

Comment:
If we move this assignment to after the for loop (and just before `checkForTermination()`) then it seems that `!terminated` would always be true (and so we wouldn't need that in `checkForTermination()`'s condition).

At that point we could make !terminated an assert (either with the `assert` keyword or `if (!terminated) throw new AssertionError()` to make sure we notice if something breaks in the future.